src/vm_cpu.rs

```diff
@@ -338,8 +340,83 @@ impl VmPmuState {
             }
         }
     }
-}
 
+    /// Saves the internal state for PMU counters. Stops started counters, and resets all configured counters.
+    /// This should be called in anticipation of an outbound context switch.
+    pub fn save_counters(&mut self) {
+        use PmuCounterState::*;
+        if let Ok(pmu_info) = pmu::PmuInfo::get() {
+            for (i, state) in self.counter_state.iter_mut().enumerate() {
+                if matches!(state, Configured(_) | Started(_)) {
+                    // Unwrap ok: Success is guaranteed
+                    let platform_counter_index = pmu_info
+                        .get_platform_counter_index(LogicalCounterIndex(i as u64))
+                        .unwrap();
+                    let result = sbi::api::pmu::stop_counters(
+                        platform_counter_index.0,
+                        0x1,
+                        PmuCounterStopFlags::default().set_reset_flag(),
+                    );
+                    if result.is_err() && !matches!(result.err(), Some(sbi::Error::AlreadyStopped))
+                    {
+                        println!("Warning: Failed to reset {platform_counter_index:?}, {result:?}");
+                        *state = Poisoned;
+                    }
+                }
+            }
+        }
+    }
+
+    /// Restores PMU counters from the saved state, reconfiguring and starting them if necessary.
+    /// This should be called in anticipation of an outbound context switch.
+    pub fn restore_counters(&mut self) {
+        use PmuCounterState::*;
+        if let Ok(pmu_info) = pmu::PmuInfo::get() {
+            for (i, state) in self.counter_state.iter_mut().enumerate() {
+                let result = match state {
```

Comment:
Might be cleaner if the body of this loop were in a helper function, e.g. `{save,restore}_one_counter`

Reply:
I see you've done this in the next commit, but it's probably cleaner if they just start out as separate helper functions to begin with.